src/sentry/api/bases/sentryapps.py

```diff
@@ -1,70 +1,226 @@
 from __future__ import absolute_import
 
-from sentry.api.authentication import ClientIdSecretAuthentication
-from sentry.api.base import Endpoint, SessionAuthentication
-from sentry.api.exceptions import ResourceDoesNotExist
-from sentry.api.permissions import ScopedPermission
-from sentry.models import SentryApp, SentryAppInstallation
+from django.http import Http404
+
 from sentry.utils.sdk import configure_scope
+from sentry.api.authentication import ClientIdSecretAuthentication
+from sentry.api.base import Endpoint
+from sentry.api.permissions import SentryPermission
+from sentry.models import SentryApp, SentryAppInstallation, Organization
+
+
+def ensure_scoped_permission(request, allowed_scopes):
+    """
+    Verifies the User making the request has at least one required scope for
+    the endpoint being requested.
+
+    If no scopes were specified in a ``scope_map``, it means the endpoint should
+    not be accessible. That is, this function expects every accessible endpoint
+    to have a list of scopes.
+
+    That list of scopes may be empty, implying that the User does not need any
+    specific scope and the endpoint is public.
+    """
+    # If no scopes were found at all, the endpoint should not be accessible.
+    if allowed_scopes is None:
+        return False
+
+    # If there are no scopes listed, it implies a public endpoint.
+    if len(allowed_scopes) == 0:
+        return True
+
+    return any(request.access.has_scope(s) for s in set(allowed_scopes))
+
+
+class SentryAppsPermission(SentryPermission):
+    scope_map = {
+        'GET': (),  # Public endoint.
+        'POST': ('org:read', 'org:integrations', 'org:write', 'org:admin'),
+    }
+
+    def has_object_permission(self, request, view, organization):
+        self.determine_access(request, organization)
+
+        if request.user.is_superuser:
+            return True
 
+        # User must be a part of the Org they're trying to create the app in.
+        if organization not in request.user.get_orgs():
+            raise Http404
+
+        return ensure_scoped_permission(
+            request,
+            self.scope_map.get(request.method),
+        )
+
+
+class SentryAppsBaseEndpoint(Endpoint):
+    permission_classes = (SentryAppsPermission, )
+
+    def convert_args(self, request, *args, **kwargs):
+        # Only the POST endpoint includes an Organization.
```

Comment:
Can we make the comment here maybe a little more descriptive, even though you went thru it with me I got confused again on https://github.com/getsentry/sentry/pull/10265/files?utf8=%E2%9C%93&diff=split#diff-7be9ee4c221344b6b849dc9a27929075R48.